### DISCUSSION TITLE
Implement "Favorites" in the menu

### DISCUSSION THREAD

**Discussion by stefonarch** [lxqt/lxqt#2264] [Ideas]
The desire  to have favorites in the menu is coming up regularly. Manually it can be achieved with
`~/.local/share/lxqt-applications.menu`:

```
	<!-- Favorites -->
	<Menu>
		<Name>Favorites</Name>
		<Directory>lxqt-favorites.directory</Directory>
		<Include><Filename>firefox.desktop</Filename></Include>
		<Include><Filename>vlc.desktop</Filename></Include>
		<Include><Filename>thunderbird.desktop</Filename></Include>
		<Include><Filename>featherpad.desktop</Filename></Include>
	</Menu>
	<!-- End Favorites -->

	<Layout>
		<Menuname show_empty="false" inline="true">LXQt-About</Menuname>
		<Separator/>
		<Merge type="files"/>
		<Merge type="menus"/>
		<Separator/>
		<Menuname>DesktopSettings</Menuname>
		<Menuname>LXQtSettings</Menuname>
		<Separator/>
		<Menuname  show_empty="true" inline="true">Favorites</Menuname>
		<Separator/>
		<Menuname show_empty="false" inline="true">X-Leave</Menuname>
		<Menuname show_empty="false" inline="true">Screensaver</Menuname>
	</Layout>

```

`~/.local/share/desktop-directories/lxqt-favorites.directory`:
```
[Desktop Entry]
Name=Favorites
...
Name[it]=Favoriti
....
Icon=favorites
Type=Directory

```
![schermata-5 lug 12:12](https://user-images.githubusercontent.com/10681413/177307086-516618c1-fdce-4297-b8da-dbb4238c6b53.png)

I'm wondering if it's possible to have in the right click menu an "add to favorites" similar as "add to desktop", using maybe  a dedicated menu file to include in the menu.

Another idea I had would be to have a "custom menu" plugin - similar to quicklaunch but which would open a list of desktop files with text. @slidinghotdog ?

**Comment by tsujan** [comment]
> The desire to have favorites in the menu is coming up regularly

Isn't that just because of habits and not practicality?

Change "favorite" to "quick". We have Quick Launch + the possibility of having multiple panels + auto-hiding on overlapping. How may favorite launchers may a user want to have? If it's 20, "favorites" will lose its sense.

**Reply by stefonarch** [reply]
I hardly use the menu, I use autostart, shortcuts and runner for launching apps. 
The menu for me is kind of a catalogue of installed apps. But people have their habits.
 I tried once the quicklaunch on a autohiding panel but it can display only icons and I am icon-blind and prefer to have text which I recognise much faster. So no docks never :D

**Reply by tsujan** [reply]
> I am icon-blind and

Oh, I hadn't considered that factor.

I'm text blind ;)

**Comment by elviosak** [comment]
I'm not too familiar with .menu files, is there sone naming/location that gets auto merged  or does it have to be a complete single file? Would we need to copy existing one, add Favorites, then select on "mainmenu" config?

Maybe the quicklaunch could be changed to have the option to act as a menu, but i'll have to look into it later, if it can be done without too many changes i'll send a PR.

I made the new custom menu a while ago as a separate plugin, and i have been using it but it's still missing some functionality and the code is kinda bad so i was going to rewrite it but have been putting it off, i'll upload to github later, but here's how it looks: https://imgur.com/n5RMvla.png
That favorites category gets saved on its own config, so multiple menus would have different favorites.

**Reply by stefonarch** [reply]
That doesn't look bad at all!

**Reply by tsujan** [reply]
@slidinghotdog 
There have been many requests for a menu like yours. It'll be great if you could polish its code and make a PR (it doesn't need to be complete for a PR, IMHO). It'll also solve the "Favorites" problem.

**Reply by elviosak** [reply]
@stefonarch @tsujan here it is, https://github.com/slidinghotdog/plugin-wingmenu these last few days i was fixing some stuff and adding keyboard navigation (not sure how good it is but IMO it's pretty intuitive) i'll leave it there for the time being, still thinking how to "clean up" the code, but when i do, i'll add it to `lxqt-panel` and send a PR. If u can, check it out and give some feedback on UI or code. Also i don't know how to edit SVG's, i think the icon is badly "extracted" (took the wing from LXQt's icon) so if someone could "fix" the icon it would be great.

**Reply by tsujan** [reply]
@slidinghotdog 
Thanks! I'd like it to be in LXQt 1.2.0 in November. The sooner you make a PR, the more time we'll have to test it and fix its probable issues.

**Reply by stefonarch** [reply]
Wow! I encountered some issues, didn't use` bash install.sh` but regular compiling script and after panel restart I get this:
![schermata-15 lug 17:30](https://user-images.githubusercontent.com/10681413/179260369-9197b94f-caa4-4098-b372-f50623bd66f0.png)
Have to test more, it looks like it made the panel crash when restarting the session. 

But yes, this _has_ to be in 1.2.0 ;)

**Reply by elviosak** [reply]
Oh i forgot to set the default menu file, it's fixed now.

**Reply by stefonarch** [reply]
Another issue: panel crashes (alert: panel crashed too many times) on session restart, but can be started fine from command line but not from lxqt-config-session's module section.

**Reply by stefonarch** [reply]
Never mind, updated and recompiled it's ok now.

**Reply by stefonarch** [reply]
BTW wing-menu is a nice name!

**Reply by stefonarch** [reply]
I can see an issue with tooltips:

https://user-images.githubusercontent.com/10681413/179342125-d5bdf73f-5f5f-4f63-a33b-e1fa7b2b5380.mp4

**Reply by elviosak** [reply]
Ok i blocked tooltips from category buttons since the information is redundant.

**Reply by stefonarch** [reply]
I had that idea too, didn't know if partially blocking tooltips was possible. On debian/intel PC I hardly can trigger those glitches.

**Comment by stefonarch** [comment]
FYI https://twitter.com/lxqt_project/status/1544434271745605634

**Comment by tsimonq2** [comment]
Darnit @stefonarch you got to it first, I'll RT from LubuntuOfficial :)

Very welcome improvement! Nice work @slidinghotdog :) you have several people in Lubuntu interested now too...

**Reply by stefonarch** [reply]
I looked at the styling: Menu size depends on icon size I learned (maybe only  1/2 or 2/3 of panel's icon size?)

And it looks that we have in all themes  blocks like this:
```
QMenu QToolButton {
    background-color: #003a55;
    padding: 5px;
    border: none;
    color: #53bcec;
    font: bold italic;
}
```
which can ruin the "categories" section (here: arch-colors) quite effective. 
I didn't find any reason to exist  for those elsewhere, so they could be removed or adapted. @tsujan ?

**Reply by tsujan** [reply]
If I remember correctly, it's for "menu titles". If you right click a plugin on the panel, you'll see two titles in the context menu: one for the plugin, the other for the panel. Those titles are really texts of (toggled) tool-buttons.

**Reply by stefonarch** [reply]
You do, I remember now. That's the reason for the `bold` and sometimes the `italic`. But I see those titles remain bold also when the whole block is removed.

**Reply by stefonarch** [reply]
`#WingMenuPlugin QToolButton { } `is working for  override I see.

**Reply by stefonarch** [reply]
https://github.com/lxqt/lxqt-themes/pull/75

Needs polishing and testing. Removed "preferences" in the` lxqt-applications-simple.menu` so there is only the config center. Downside of that is that the search does'nt find "appearance" and so  on. "Valendas" theme in this screen.

![screenshot-533d9e6a](https://user-images.githubusercontent.com/10681413/179427381-b319fc92-8d35-4e24-94aa-bef96849796c.png)
.

**Comment by stefonarch** [comment]
@slidinghotdog any eta for a PR?

**Reply by elviosak** [reply]
@stefonarch sorry, i've been away for a while but i think i can send it in a week or two.

**Reply by stefonarch** [reply]
Cool, thanks, it would be real nice to have this in 1.2.0!
